internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java

```diff
@@ -0,0 +1,673 @@
+            getLatestOffsetWithRetries(pubSubTopicPartition, DEFAULT_MAX_RETRIES_FOR_POPULATING_TMD_CACHE_ENTRY);
+        return new ValueAndExpiryTime<>(latestOffset, now + cachedEntryTtlInNs);
+      });
+    } catch (PubSubTopicDoesNotExistException | PubSubOpTimeoutException e) {
+      LOGGER.error("Failed to get end offset for topic-partition: {}", pubSubTopicPartition, e);
+      return StatsErrorCode.LAG_MEASUREMENT_FAILURE.code;
+    }
+    populateCacheWithLatestOffset(pubSubTopicPartition);
+    return cachedValue.getValue();
+  }
+
+  // load the cache with the latest offset
+  void populateCacheWithLatestOffset(PubSubTopicPartition pubSubTopicPartition) {
+    long now = System.nanoTime();
+    // if there is no entry in the cache or the entry has expired, we will try to populate the cache
+    ValueAndExpiryTime<Long> cachedValue = latestOffsetCache.putIfAbsent(
+        pubSubTopicPartition,
+        new ValueAndExpiryTime<>((long) StatsErrorCode.LAG_MEASUREMENT_FAILURE.code, now + cachedEntryTtlInNs));
+    if (cachedValue == null || (cachedValue.getExpiryTimeNs() <= now && cachedValue.tryAcquireUpdateLock())) {
+      CompletableFuture<Long> future =
+          getLatestOffsetWithRetriesAsync(pubSubTopicPartition, DEFAULT_MAX_RETRIES_FOR_POPULATING_TMD_CACHE_ENTRY);
+      future.whenComplete((latestOffset, throwable) -> {
+        if (throwable != null) {
+          latestOffsetCache.remove(pubSubTopicPartition);
+          return;
+        }
+        latestOffsetCache.get(pubSubTopicPartition).updateValue(latestOffset);
+      });
+    }
+  }
+
+  // API: Get the offset for a given timestamp
+  long getOffsetForTime(PubSubTopicPartition pubSubTopicPartition, long timestamp) {
+    // We start by retrieving the latest offset. If the provided timestamp is out of range,
+    // we return the latest offset. This ensures that we don't miss any records produced
+    // after the 'offsetForTime' call when the latest offset is obtained after timestamp checking.
+    long latestOffset = getLatestOffset(pubSubTopicPartition);
+
+    PubSubConsumerAdapter pubSubConsumerAdapter = acquireConsumer();
+    try {
+      long startTime = System.currentTimeMillis();
+      Long result = pubSubConsumerAdapter
+          .offsetForTime(pubSubTopicPartition, timestamp, PUBSUB_OFFSET_API_TIMEOUT_DURATION_DEFAULT_VALUE);
+      TopicManagerStats.recordLatency(
+          topicManagerStats,
+          TopicManagerStats.OCCURRENCE_LATENCY_SENSOR_TYPE.GET_OFFSET_FOR_TIME,
+          System.currentTimeMillis() - startTime);
+      if (result == null) {
+        // when offset is null, it means the given timestamp is either
+        // out of range or the topic does not have any message
+        return latestOffset;
+      }
+      return result;
+    } finally {
+      releaseConsumer(pubSubConsumerAdapter);
+    }
+  }
+
+  long getOffsetForTimeWithRetries(PubSubTopicPartition pubSubTopicPartition, long timestamp, int retries) {
```

Comment:
I would prefer to continue leveraging `RetryUtils` otherwise the retry code is very bespoke and scattered... I have noticed that the new retry code has no sleep in-between attempts (neither exponential nor fixed), which does not seem ideal to me. Even if indeed we want retries without backoff, we should be able to achieve that as well with `RetryUtils`.